### DISCUSSION TITLE
Nextjs + SSR Access Token Not Refreshing

### DISCUSSION THREAD

**Discussion by J0ELh** [supabase/supabase#26757] [Questions] [answered]
I've already started discussing this question in a previous [question](https://github.com/orgs/supabase/discussions/26718#discussion-6718991), however I have been unable to resolve the problem and I don't know what else to try. 

**Problem:** In my application which uses next js and server side rendering, I followed [this guide](https://supabase.com/docs/guides/auth/server-side/nextjs?queryGroups=router&router=app). After my access token expires, I find that I get logged out of the application (```await supabase.auth.getUser();``` and ```await supabase.auth.getSession();``` return null).

What I have tried:

- copying the supabase docs exactly
- monitoring cookies in browser
- switching browser & disabling extensions
- logging all outputs and token expiry time (which is why i know the access token is never refreshed)
- manually checking the JWT token info

Here's how my code is setup:
my src/lib/supabase/middleware.ts is supposed to refresh my access token if it's about to expire **automatically**
```
import { createServerClient, type CookieOptions } from "@supabase/ssr";
import { NextResponse, type NextRequest } from "next/server";

export async function updateSession(request: NextRequest) {
  let response = NextResponse.next({
    request: {
      headers: request.headers,
    },
  });

  const supabase = createServerClient(
    process.env.NEXT_PUBLIC_SUPABASE_URL!,
    process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY!,
    {
      cookies: {
        get(name: string) {
          return request.cookies.get(name)?.value;
        },
        set(name: string, value: string, options: CookieOptions) {
          request.cookies.set({
            name,
            value,
            ...options,
          });
          response = NextResponse.next({
            request: {
              headers: request.headers,
            },
          });
          response.cookies.set({
            name,
            value,
            ...options,
          });
        },
        remove(name: string, options: CookieOptions) {
          request.cookies.set({
            name,
            value: "",
            ...options,
          });
          response = NextResponse.next({
            request: {
              headers: request.headers,
            },
          });
          response.cookies.set({
            name,
            value: "",
            ...options,
          });
        },
      },
    }
  );

  // refreshing the auth token
  await supabase.auth.getUser();

  return response;
}
```

Any request from my client that needs my backend server, calls the /api route in order to pick up the JWT for verification. this is what that route.ts looks like:
```
import { getServerAccessToken } from "@/lib/supabase/actions";
import { NextResponse } from "next/server";

export async function POST(request: Request): Promise<Response> {
  try {
    let { path, method, headers, body, timeout } = await request.json();

    // attach auth token
    const token = await getServerAccessToken();
    headers["Authorization"] = `Bearer ${token}`;

    let options: RequestInit = {
      method,
      headers: headers as HeadersInit,
      body,
    };

    // manage custom timeout for long requests
    let timeoutId;

    if (timeout) {
      const controller = new AbortController();
      timeoutId = setTimeout(() => controller.abort(), timeout); // Set timeout
      options["signal"] = controller.signal;
    }

    const url = `${process.env.NEXT_PUBLIC_API_URL}${path}`
    const response = await fetch(url, options);

    if (timeoutId) {
      clearTimeout(timeoutId);
    }

    if (!response.ok) {
      const errorData = await response.json();

      console.error(errorData);

      return NextResponse.json(
        {
          error: errorData.detail || "Unknown error",
          status: response.status,
        },
        {
          status: response.status,
        }
      );
    }

    return response;
  } catch (error) {
    console.error("Error in API route handler:", error);
    return NextResponse.json(
      {
        error: "An unexpected error occurred",
        status: 500,
      },
      {
        status: 500,
      }
    );
  }
}
```

And finally, here's the ```getServerAccessToken``` function that is the place where the app crashes due to user and/or session being null.

```

export const getServerAccessToken = async () => {
  const supabase = createClient();

  const {
    data: { user },
  } = await supabase.auth.getUser();

  if (!user) {
    throw new Error("No valid user");
  }

  const {
    data: { session },
  } = await supabase.auth.getSession();

  if (!session) {
    throw new Error("No valid session");
  }

  return session.access_token;
};

export const googleLogin = async () => {
  const supabase = createClient();

  const origin = headers().get("origin");

  const { data, error } = await supabase.auth.signInWithOAuth({
    provider: "google",
    options: {
      redirectTo: `${process.env.NEXT_PUBLIC_CLIENT_URL}/auth/google`,
    },
  });

  if (data.url) {
    redirect(data.url); // use the redirect API for your server framework
  }
};
```

My client.ts and server.ts in the src/lib/supabase directory and the middleware.ts in the root directory look exactly like specified in the supabase docs linked above. 

Extra Code in case it helps:

login route.ts
```
import { createClient } from "@/lib/supabase/server";
import { LoginSchema } from "@/lib/types/zod";
import { formatZodErrors } from "@/lib/utils";
import { NextResponse } from "next/server";

export async function POST(request: Request) {
  const body: unknown = await request.json();

  const result = LoginSchema.safeParse(body);

  if (!result.success) {
    return NextResponse.json({ errors: formatZodErrors(result.error) });
  }

  const supabase = createClient();
  const { error } = await supabase.auth.signInWithPassword(result.data);

  if (error) {
    console.log(error);
    return NextResponse.json({ errors: { supabase: error.message } });
  }

  const { searchParams, origin } = new URL(request.url);
  return NextResponse.redirect(`${process.env.NEXT_PUBLIC_CLIENT_URL}/dashboard`, 303);
}
```

**Comment by Yassdrk** [comment]
Hello,
Same issue with same middleware, any tips ?

Regards

**Reply by J0ELh** [reply, marked as answer]
When I started monitoring the cookie usage, I realized that they were not being refreshed. So, when my access token expired, I **did** receive a new access token from my refresh token. However this was not saved to my cookies and that's why I was logged out. The solution was to reset all middleware, and supabase related files back to exactly what they look like in the docs. There were some other changes I needed to make, but that was the largest change. Hope that helps!

**Reply by Yassdrk** [reply]
> When I started monitoring the cookie usage, I realized that they were not being refreshed. So, when my access token expired, I **did** receive a new access token from my refresh token. However this was not saved to my cookies and that's why I was logged out. The solution was to reset all middleware, and supabase related files back to exactly what they look like in the docs. There were some other changes I needed to make, but that was the largest change. Hope that helps!

I just did the same and i think it works too ! :)

**Reply by firstian** [reply]
Can you guys tell me whether your cookies are split into chunks? What I found is that the refresh happens, but only the last chunk got updated during a refresh. I look at the sample code, and I wonder if there is a bug there?

**Reply by ivanbogdan-git** [reply]
Hi @J0ELh, any way you could share the code for this? 

I'm constantly getting 
```
AuthApiError: Invalid Refresh Token: Session Expired
```
as an error in the Vercel logs, and I don't handle auth on browser client at all. Could you share your middleware code for this to work properly?

**Reply by J0ELh** [reply]
> Hi @J0ELh, any way you could share the code for this?
> 
> I'm constantly getting
> 
> ```
> AuthApiError: Invalid Refresh Token: Session Expired
> ```
> 
> as an error in the Vercel logs, and I don't handle auth on browser client at all. Could you share your middleware code for this to work properly?

Sorry @ivanbogdan-git, I no longer have access to the project I was working on at the time. All I can say is that it was painful to debug but what worked was going back to follow the supabase documentation precisely.

**Reply by dule-kitemetric** [reply]
@ivanbogdan-git isn't it fixed? How did you do that?
I'm facing the same problem.

**Reply by ivanbogdan-git** [reply]
@dule-kitemetric I contacted the auth support, they helped me fix it, i'm not entirely sure if i still remember and have access to repo/supabase acc.
